homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,126 @@
-
-            return None
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve current app for %s", self.device_id)
-            return None
-
-    @property
-    def running_apps(self):
-        """Return a list of running apps."""
-        try:
-            response = requests.get(
-                DEVICE_APPS_URL.format(
-                    self.proto, self.host, self.port, self.device_id, 'running'
-                    ), timeout=10).json()
-            return response.get('running_apps')
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not retrieve running apps for %s", self.device_id)
-            return None
-
-    def action(self, action_id):
-        """Perform an action on the device."""
-        try:
-            requests.get(DEVICE_ACTION_URL.format(
-                self.proto, self.host, self.port, self.device_id, action_id
-                ), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Action request for %s was not accepted for device %s",
-                action_id, self.device_id)
-
-    def start_app(self, app_name):
-        """Start an app."""
-        try:
-            requests.get(DEVICE_APPS_URL.format(
-                self.proto, self.host, self.port, self.device_id,
-                app_name + '/start'), timeout=10)
-        except requests.exceptions.RequestException:
-            _LOGGER.error(
-                "Could not start %s on %s", app_name, self.device_id)
+    adbkey = config.get(CONF_ADBKEY)
+    get_source = config.get(CONF_GET_SOURCE)
+    get_sources = config.get(CONF_GET_SOURCES)
+
+    device = FireTVDevice(host, name, adbkey, get_source, get_sources)
+    adb_log = " using adbkey='{0}'".format(adbkey) if adbkey else ""
+    if not device._firetv._adb:
+        _LOGGER.warning("Could not connect to Fire TV at %s%s", host, adb_log)
+
+        # Debugging
+        if adbkey != "":
+            # Check whether the key files exist
+            if not os.path.exists(adbkey):
+                raise FileNotFoundError(
+                    "ADB private key {} does not exist".format(adbkey))
+            if not os.path.exists(adbkey + ".pub"):
+                raise FileNotFoundError(
+                    "ADB public key {} does not exist".format(adbkey + '.pub'))
+
+            # Check whether the key files can be read
+            with open(adbkey):
+                pass
+            with open(adbkey + '.pub'):
+                pass
+
+    else:
+        _LOGGER.info("Setup Fire TV at %s%s", host, adb_log)
+        add_devices([device])
+
+
+def adb_wrapper(func):
+    """A wrapper that will wait if previous ADB commands haven't finished."""
+    @functools.wraps(func)
+    def _adb_wrapper(self, *args, **kwargs):
+        attempts = 0
+        while self._adb_lock and attempts < 5:
```

Comment:
Accessing a protected variable.

Reply:
Fixed in https://github.com/JeffLIrion/home-assistant/commit/91b7b9b73ae68dd2c0cf087a53d3efc1a89258be (requires firetv 1.0.7 to be released on pypi)